### PR TITLE
Add unicode letters generator

### LINKED ISSUE #69: Random UTF-8 string generation must generate valid string
FauxFactory have generated the following string:

```python
u'\u50dd\uf3b6\u1365\u4e5c\u2441\ue94f\u6b18\u4921\u6fc3\u052c'
```

But `\ue94f` [1] and `\uf3b6` [2] are not valid unicode characters. FauxFactory need a better way to generate random unicode strings.

Thanks @Ichimonji10 for helping in the investigation process.

[1] http://www.fileformat.info/info/unicode/char/e94f/index.htm
[2] http://www.fileformat.info/info/unicode/char/f3b6/index.htm

### PR DESCRIPTION
**Pull request by elyezer**
This generator is a helper for the gen_utf8 function which will provide
the system supported list of unicode letters. This will avoid generating
unicode string with control characters and other non letters characters.

Also adds tests for the generator in order to ensure it is not
generating unwanted characters.

Closes #69

### CODE REVIEW

File: fauxfactory/__init__.py [hunk 1]
```diff
@@ -72,6 +73,29 @@ def _is_positive_int(length):
...
 
+def _unicode_letters_generator():
+    """Generates unicode characters in the letters category
+
+    :return: a generator which will generates all unicode letters available
+
+    """
+    if sys.version_info.major == 2:
+        chr_function = unichr
+        range_function = xrange
```
Comment (Ichimonji10) [inline fauxfactory/__init__.py]: chr_function = unichr # pylint:disable=undefined-variable
    range_function = xrange # pylint:disable=undefined-variable

File: fauxfactory/__init__.py [hunk 2]
```diff
@@ -72,6 +73,29 @@ def _is_positive_int(length):
...
+    else:
+        range_function = range
+        chr_function = chr
```
Comment (Ichimonji10) [inline fauxfactory/__init__.py]: Can you order the two variable assignments the same way in both the `if` and `else`? It's a minor thing, but it helps me quickly compare the two blocks of code.

File: fauxfactory/__init__.py [hunk 3]
```diff
@@ -673,33 +697,9 @@ def gen_utf8(length=10):
...
-
-    # Convert codepoints to characters. Python 2 and 3 support the `unichr`
-    # and `chr` functions, respectively.
-    if sys.version_info.major == 2:
-        # pylint:disable=E0602
-        output = u''.join(unichr(codepoint) for codepoint in codepoints)
-    else:
-        output = u''.join(chr(codepoint) for codepoint in codepoints)
-    return _make_unicode(output)
+    return _make_unicode(u''.join(
```
Comment (Ichimonji10) [inline fauxfactory/__init__.py]: Is the call to `_make_unicode` necessary here, given that the string being joined is `u''`?

**Comment by Ichimonji10** [comment]
ACK pending comments

File: fauxfactory/__init__.py (hunk 2 above)
Comment (elyezer) [inline fauxfactory/__init__.py]: Sure, thanks for catching this.

File: fauxfactory/__init__.py (hunk 3 above)
Comment (elyezer) [inline fauxfactory/__init__.py]: It is not required, thank you again.

**Comment by Ichimonji10** [comment]
[5.5.1 General Category Values](http://www.unicode.org/reports/tr44/tr44-4.html#Property_Values) explains the meaning of prefixes such as "Lu" or "No".

**Comment by Ichimonji10** [comment]
Do you think the types of characters emitted by `gen_utf8` is appropriate? Character categories such as "Decimal_Number" and "Currency_Symbol" are omitted. And, as Ke$ha demonstrates, those values actually are used in some weird places. We can make a judgment call and omit them. But in that case, the docstring for `gen_utf8` should be updated appropriately.

**Comment by elyezer** [comment]
@Ichimonji10 I have updated the docstring. For testing purposes I think just the letters should suffice as probably we will get at least one UTF-8 glyph with 2 or more bytes. We can add more categories for sure, but I think is better to "stay safe", just using a multi-byte character will define if the system is capable of handling UTF-8 strings.

**Comment by elyezer** [comment]
On my system I got a total of 48270 unicode letters. It is 73.66% of the maxunicode 65535.

**Comment by Ichimonji10** [comment]
ACK

**Comment by elyezer** [comment]
Have done some practical testing and here are the results:

With this changes:

```
$ py.test tests/foreman/api/test_user.py -k test_create
================================ test session starts ================================
platform darwin -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
collected 23 items

tests/foreman/api/test_user.py ...

====================== 20 tests deselected by '-ktest_create' =======================
===================== 3 passed, 20 deselected in 13.24 seconds ======================

$ py.test tests/foreman/api/test_user.py -k test_create
================================ test session starts ================================
platform darwin -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
collected 23 items

tests/foreman/api/test_user.py ...

====================== 20 tests deselected by '-ktest_create' =======================
====================== 3 passed, 20 deselected in 7.42 seconds ======================

$ py.test tests/foreman/api/test_user.py -k test_create
================================ test session starts ================================
platform darwin -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
collected 23 items

tests/foreman/api/test_user.py ...

====================== 20 tests deselected by '-ktest_create' =======================
====================== 3 passed, 20 deselected in 8.17 seconds ======================

$ py.test tests/foreman/api/test_user.py -k test_create
================================ test session starts ================================
platform darwin -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
collected 23 items

tests/foreman/api/test_user.py ...

====================== 20 tests deselected by '-ktest_create' =======================
===================== 3 passed, 20 deselected in 13.62 seconds ======================

 $ py.test tests/foreman/api/test_user.py -k test_create
================================ test session starts ================================
platform darwin -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
collected 23 items

tests/foreman/api/test_user.py ...

====================== 20 tests deselected by '-ktest_create' =======================
===================== 3 passed, 20 deselected in 28.24 seconds ======================
```

Without the changes

```
$ py.test tests/foreman/api/test_user.py -k test_create
================================ test session starts ================================
platform darwin -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
collected 23 items

tests/foreman/api/test_user.py FFF

===================================== FAILURES ======================================
 UsersTestCase.test_create_1__u_firstname___u__u28ff_u5d9f_ueca0_u2890_ub79b_u4a90_u2e59_u461b_u31ac__

------------------------------- Captured stdout call --------------------------------
2015-02-15 11:39:37 - nailgun.client - WARNING - Received HTTP 422 response: {
  "error": {"id":null,"errors":{"firstname":["is invalid"]},"full_messages":["First name is invalid"]}
}

2015-02-15 11:39:38 - nailgun.client - WARNING - Received HTTP 422 response: {
  "error": {"id":null,"errors":{"lastname":["is invalid"]},"full_messages":["Surname is invalid"]}
}

2015-02-15 11:39:39 - nailgun.client - WARNING - Received HTTP 422 response: {
  "error": {"id":null,"errors":{"login":["is invalid"]},"full_messages":["Username is invalid"]}
}

====================== 20 tests deselected by '-ktest_create' =======================
====================== 3 failed, 20 deselected in 4.93 seconds ======================

$ py.test tests/foreman/api/test_user.py -k test_create
================================ test session starts ================================
platform darwin -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
collected 23 items

tests/foreman/api/test_user.py FFF

===================================== FAILURES ======================================

2015-02-15 11:39:42 - nailgun.client - WARNING - Received HTTP 422 response: {
  "error": {"id":null,"errors":{"firstname":["is invalid"]},"full_messages":["First name is invalid"]}
}

2015-02-15 11:39:43 - nailgun.client - WARNING - Received HTTP 422 response: {
  "error": {"id":null,"errors":{"lastname":["is invalid"]},"full_messages":["Surname is invalid"]}
}

2015-02-15 11:39:44 - nailgun.client - WARNING - Received HTTP 422 response: {
  "error": {"id":null,"errors":{"login":["is invalid"]},"full_messages":["Username is invalid"]}
}

====================== 20 tests deselected by '-ktest_create' =======================
====================== 3 failed, 20 deselected in 3.81 seconds ======================

$ py.test tests/foreman/api/test_user.py -k test_create
================================ test session starts ================================
platform darwin -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
collected 23 items

tests/foreman/api/test_user.py FFF

===================================== FAILURES ======================================

2015-02-15 11:39:47 - nailgun.client - WARNING - Received HTTP 422 response: {
  "error": {"id":null,"errors":{"firstname":["is invalid"]},"full_messages":["First name is invalid"]}
}

2015-02-15 11:39:48 - nailgun.client - WARNING - Received HTTP 422 response: {
  "error": {"id":null,"errors":{"lastname":["is invalid"]},"full_messages":["Surname is invalid"]}
}

2015-02-15 11:39:49 - nailgun.client - WARNING - Received HTTP 422 response: {
  "error": {"id":null,"errors":{"login":["is invalid"]},"full_messages":["Username is invalid"]}
}

====================== 20 tests deselected by '-ktest_create' =======================
====================== 3 failed, 20 deselected in 3.47 seconds ======================
```

For this test I commented out all non UTF-8 test data and I removed most of the failures output and left just the invalid messages.

**Comment by omaciel** [comment]
ACK

**Pull request merged**